LongestPole/Rectangles.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2024 Lean FRO. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Morrison
+-/
+
+/-!
+# Finding the maximal rectangles in a list of points.
+-/
+
+structure Rectangle where
+  left : Nat
+  right : Nat
+  bottom : Nat
+  top : Nat
+deriving DecidableEq, Repr
+
+namespace Rectangle
+
+def width (r : Rectangle) : Nat := r.right - r.left
+def height (r : Rectangle) : Nat := r.top - r.bottom
+
+def toString (r : Rectangle) : String :=
+  s!"[{r.left}, {r.right}) × [{r.bottom}, {r.top})"
+
+instance : ToString Rectangle where
+  toString := toString
+
+instance instPointMembership : Membership (Nat × Nat) Rectangle where
+  mem r x := r.left ≤ x.1 ∧ x.1 < r.right ∧ r.bottom ≤ x.2 ∧ x.2 < r.top
+
+instance (x : Nat × Nat) (r : Rectangle) : Decidable (x ∈ r) := by
+  dsimp [Membership.mem]
+  infer_instance
+
+instance instLE : LE Rectangle where
+  le r₁ r₂ := r₁.width = 0 ∨ r₁.height = 0 ∨ ((r₁.left, r₁.bottom) ∈ r₂ ∧ (r₁.right - 1, r₁.top - 1) ∈ r₂)
+
+instance : DecidableRel ((· : Rectangle) ≤ ·) := by
+  dsimp [LE.le, instLE]
+  infer_instance
+
+instance : Membership Rectangle (List (Nat × Nat)) where
+  mem xs r := (List.range' r.left r.width).all fun x => (List.range' r.bottom r.height).all fun y => (x, y) ∈ xs
+
+instance (r : Rectangle) (xs : List (Nat × Nat)) : Decidable (r ∈ xs) := by
+  dsimp [Membership.mem]
+  infer_instance
+
+def expansions (r : Rectangle) : List Rectangle :=
```

Comment:
```suggestion
/--
Given a rectangle `r`, return the up to four four rectangles obtained from `r` by
- shifting the bottom left corner one coordinate to the left resp. bottom (if possible), or
- shifting the top right corner one coordinate to the right resp. top.
-/

Reply:
If you want to prove a lemma: `r` is contained in each of its expansions.

Reply:
again your suggestion removes the `def` line...

Reply:
No lemmas, please, we're programming. :-)